### PR TITLE
UI-281 Update schema:check output format

### PR DESCRIPTION
**Pull request by justinanastos**
We use `Listr` to output the progress list (✓s and ✖s). We must pre-fill the titles of the steps that have yet to run. I then change those titles into the correct output. I'd love some feedback on better titles:

```
  ✔ Loading Apollo Project
  ⠦ Validating local schema against tag prod on service engine
    → Validating schema
    Comparing schema changes
    Reporting result
```

---

TODO:

- [x] Update `CHANGELOG.md` with your change (include reference to issue & this PR)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

---

- [x] Do some more testing on reporting with no errors
- [x] Add URL at the bottom

### CODE REVIEW

File: packages/apollo/src/commands/service/check.ts [hunk 1]
```diff
@@ -193,15 +357,17 @@ export default class ServiceCheck extends ProjectCommand {
...
-      diffToPrevious: { changes, validationConfig }
-    } = checkSchemaResult;
-    const failures = changes.filter(({ type }) => type === ChangeType.FAILURE);
-
     if (shouldOutputJson) {
       return this.log(
-        JSON.stringify({ targetUrl, changes, validationConfig }, null, 2)
+        JSON.stringify(
+          {
+            targetUrl: checkSchemaResult,
```
Comment (trevor-scheer) [inline packages/apollo/src/commands/service/check.ts]: Is this line correct? Looks like it's getting everything instead of just a relevant slice of the data.

File: CHANGELOG.md [hunk 2]
```diff
@@ -37,6 +37,7 @@
   - Include targetUrl in the output of the `service:check` command [#1072](https://github.com/apollographql/apollo-tooling/pull/1072)
   - Import apollo-env utility types directly instead of treating them as globals [#1074](https://github.com/apollographql/apollo-tooling/pull/1074)
   - Add `--markdown` output option to `service:check` [#1072](https://github.com/apollographql/apollo-tooling/pull/1072)
+  - Enhance formatting for `service:check` output [#1146](https://github.com/apollographql/apollo-tooling/pull/1146)
```
Comment (trevor-scheer) [inline CHANGELOG.md]: This entry and the one above it (I believe) should be placed under the Upcoming header, rather than within an already published version of the changelog. Ping me if you're unsure!

**Comment by trevor-scheer** [comment]
Looks like some funny business on windows vs. linux shell output based on the snapshot failures. Not sure what the right thing to do here is, but happy to talk about it tomorrow 👍

File: packages/apollo/src/commands/service/__tests__/check.test.ts [hunk 3]
```diff
@@ -1,52 +1,159 @@
...
+🔢 Compared **0 schema changes** against operations seen over the **last 24 hours**.
 ✅ Found **no breaking changes**.
 
 🔗 [View your service check details](https://engine-dev.apollographql.com/service/engine/checks?schemaTag=Detached%3A%20d664f715645c5f0bb5ad4f2260cd6cb8d19bbc68&schemaTagId=f9f68e7e-1b5f-4eab-a3da-1fd8cd681111&from=2019-03-26T22%3A25%3A12.887Z).
 "
 `);
+    });
+  });
+
+  describe("formatTimePeriod", () => {
```
[the remark this answers is not in the document]
Comment (justinanastos) [reply packages/apollo/src/commands/service/__tests__/check.test.ts]: Great advice. The 1 hour code is wrong :)

File: packages/apollo/src/commands/service/check.ts [hunk 4]
```diff
@@ -97,6 +116,59 @@ ${
 `;
 }
 
+export function formatHumanReadable({
+  checkSchemaResult
+}: {
+  checkSchemaResult: CheckSchema_service_checkSchema;
+}): string {
```
[the remark this answers is not in the document]
Comment (justinanastos) [reply packages/apollo/src/commands/service/check.ts]: Yes, we can let TS infer the return type. I defined it as a check to myself to make sure I always return a string and not something else (even if it's inferred) by accident. If it's a common practice to infer the return type then we should stick with it.

File: CHANGELOG.md (hunk 2 above)
Comment (justinanastos) [reply CHANGELOG.md]: oh man what a silly mistake

File: packages/apollo/src/commands/service/check.ts [hunk 5]
```diff
@@ -131,56 +203,148 @@ export default class ServiceCheck extends ProjectCommand {
...
+              task.title = `Compared ${chalk.blue(
+                schemaChanges.length.toString()
+              )} schema ${
+                schemaChanges.length === 1 ? "change" : "changes"
+              } against ${chalk.blue(numberOfCheckedOperations.toString())} ${
+                numberOfCheckedOperations === 1 ? "operation" : "operations"
+              }${
+                days
+                  ? ` seen in the past ${chalk.blue(
+                      days === 1 ? "24 hours" : `${days} days`
```
[the remark this answers is not in the document]
Comment (justinanastos) [reply packages/apollo/src/commands/service/check.ts]: Great point. I changed this to reuse `formatTimePeriod`.

File: packages/apollo/src/commands/service/check.ts (hunk 1 above)
Comment (justinanastos) [reply packages/apollo/src/commands/service/check.ts]: Great catch!

File: packages/apollo/src/commands/service/check.ts [hunk 6]
```diff
@@ -221,25 +387,15 @@ export default class ServiceCheck extends ProjectCommand {
...
+    this.log(formatHumanReadable({ checkSchemaResult }));
+
     // exit with failing status if we have failures
-    if (failures.length > 0) {
+    if (
+      checkSchemaResult.diffToPrevious.changes.find(
+        ({ type }) => type === ChangeType.FAILURE
+      )
+    ) {
       this.exit();
```
[the remark this answers is not in the document]
Comment (justinanastos) [reply packages/apollo/src/commands/service/check.ts]: Agreed. Should update now!

File: packages/apollo/src/commands/service/check.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply packages/apollo/src/commands/service/check.ts]: Totally ok with leaving as-is! My default has always been to let it infer unless we need to tell it otherwise, but I appreciate your reason and am reconsidering my default.

File: CHANGELOG.md (hunk 2 above)
Comment (trevor-scheer) [reply CHANGELOG.md]: There's literally no documentation about how to update the changelog, so I don't think it's silly.

File: packages/apollo/src/commands/service/check.ts [hunk 7]
```diff
@@ -131,15 +131,22 @@ export function formatHumanReadable({
   if (changes.length === 0) {
     result = "\nNo changes present between schemas";
   } else {
-    const breakingChanges = changes.filter(
+    // Create a sorted list of the changes. We'll then filter values from the sorted list, resulting in sorted
+    // filtered lists.
+    const sortedChanges = sortBy<typeof changes[0]>(changes, [
```
Comment (evans) [inline packages/apollo/src/commands/service/check.ts]: Does `sortBy` infer the type of `changes`?

I thought it would 😞

File: packages/apollo/src/commands/service/__tests__/check.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply packages/apollo/src/commands/service/__tests__/check.test.ts]: I actually think it _wasn't_ wrong, but the change you made is also not incorrect. I only asked for the test since I could see the edge case in the function, but it made sense in the way that you would use it.

"xxx in the last hour"
"xxx in the last 24 hours"

So I think you've changed it to
"xxx in the last 1 hour"

Either way seems fine to me! 👍

File: packages/apollo/src/commands/service/check.ts (hunk 7 above)
Comment (justinanastos) [reply packages/apollo/src/commands/service/check.ts]: Unfortunately no; it appears as though the array syntax causes confusion.

File: packages/apollo/src/commands/service/__tests__/check.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinanastos) [reply packages/apollo/src/commands/service/__tests__/check.test.ts]: Good eye 😃 You're right!
